tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -117,50 +117,82 @@ impl ServerKey {
-        shift: usize,
+        shift: u64,
     ) {
-        let tmp = self.key.message_modulus.0 as f64;
-
-        //number of bits of message
-        let nb_bits = tmp.log2() as usize;
+        // see parallel implementation for a bit more details
 
-        // 2^u = 2^{p*q+r} = 2^{p*(q+1)}*2^{r-p}
-        let quotient = shift / nb_bits;
+        debug_assert!(ct.block_carries_are_empty());
+        debug_assert!(self.key.carry_modulus.0 >= self.key.message_modulus.0 / 2);
 
-        //p-r
-        let modified_remainder = nb_bits - (shift % nb_bits);
+        if shift == 0 {
+            return;
+        }
 
-        //if r == 0
-        if modified_remainder == nb_bits {
-            self.full_propagate(ct);
-            self.blockshift_right_assign(ct, quotient);
-        } else {
-            // B/2^u = (B*2^{p-r}) / (2^{p*(q+1)})
-            self.unchecked_scalar_left_shift_assign(ct, modified_remainder);
+        let num_bits_in_block = (self.key.message_modulus.0 as f64).log2() as u64;
+        let rotations = ((shift / num_bits_in_block) as usize).min(ct.blocks.len());
+        let shift_within_block = shift % num_bits_in_block;
+        let num_blocks = ct.blocks.len();
 
-            // We partially propagate in order to not lose information
-            self.partial_propagate(ct);
-            self.blockshift_right_assign(ct, 1_usize);
+        // rotate left as the blocks are from LSB to MSB
+        ct.blocks.rotate_left(rotations);
+        for block in &mut ct.blocks[num_blocks - rotations..] {
+            *block = self.key.create_trivial(0);
+        }
 
-            // We propagate the last block in order to not lose information
-            self.propagate(ct, ct.blocks.len() - 2);
-            self.blockshift_right_assign(ct, quotient);
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
         }
-    }
 
-    /// Propagates all carries except the last one.
-    /// For development purpose only.
-    fn partial_propagate<PBSOrder: PBSOrderMarker>(&self, ctxt: &mut RadixCiphertext<PBSOrder>) {
-        let len = ctxt.blocks.len() - 1;
-        for i in 0..len {
-            self.propagate(ctxt, i);
+        let message_modulus = self.key.message_modulus.0 as u64;
+        let lut =
+            self.key
+                .generate_accumulator_bivariate(|mut current_block, mut previous_block| {
+                    // left shift not to lose
+                    // bits when shifting right afterwards
+                    current_block <<= num_bits_in_block;
+                    previous_block <<= num_bits_in_block;
+
+                    current_block >>= shift_within_block;
+                    previous_block >>= shift_within_block;
+
+                    // The way of gettint carry / message is reversed compared
+                    // to the usual way but its normal
+                    let message_of_current_block = current_block >> num_bits_in_block;
```

Comment:
no it wont overflow the full code is

```                    // left shift not to lose
                    // bits when shifting right afterwards
                    current_block <<= num_bits_in_block;
                    previous_block <<= num_bits_in_block;

                    current_block >>= shift_within_block;
                    previous_block >>= shift_within_block;
```

Where current_block and previous_block are in [0...message_modulus-1] 
and shift_within_block < num_bits_in_block, because shift_within_block  = shift % num_bits_in_block

The we first left shift by a value that is strictly greater than the value we are right shifting afterwards

Reply:
yeah but here you re shift by num_bits_in_block right ?

Reply:
I agree on the good result, I just wonder if we should prefer a shift which underflows or the division the same way you did for the other scalar shift :)

Reply:
Visualizing shift works better in my brain than division, and the underflow here is defined behaviour

Reply:
then all good